0000-spec-distribution.md

```diff
@@ -0,0 +1,313 @@
+
+### Execution Order
+
+During detection, a buildpack ID or list of buildpack IDs is resolved into individual buildpack implementation that include a `path` field.
+
+First, the detector determines the user's choice of buildpack IDs or builder's order definition.
+
+Next, the 2-D ordering of buildpacks is derived as follows:
+
+Where:
+- O and P are buildpacks IDs referencing buildpacks that compose other buildpack IDs.
+- A through H are buildpack IDs referencing executable buildpacks. 
+
+Given:
+
+<img src="http://tex.s2cms.ru/svg/%0AO%20%3D%0A%5Cbegin%7Bbmatrix%7D%0AA%2C%20%26%20B%20%5C%5C%0AC%2C%20%26%20D%0A%5Cend%7Bbmatrix%7D%0A" alt="
+O =
+\begin{bmatrix}
+A, &amp; B \\
+C, &amp; D
+\end{bmatrix}
+" />
+
+<img src="http://tex.s2cms.ru/svg/%0AP%20%3D%0A%5Cbegin%7Bbmatrix%7D%0AE%2C%20%26%20F%20%5C%5C%0AG%2C%20%26%20H%0A%5Cend%7Bbmatrix%7D%0A" alt="
+P =
+\begin{bmatrix}
+E, &amp; F \\
+G, &amp; H
+\end{bmatrix}
+" />
+
+We propose:
+
+<img src="http://tex.s2cms.ru/svg/%0A%5Cbegin%7Bbmatrix%7D%0AE%2C%20%26%20O%2C%20%26%20F%0A%5Cend%7Bbmatrix%7D%20%3D%20%0A%5Cbegin%7Bbmatrix%7D%0AE%2C%20%26%20A%2C%20%26%20B%2C%20%26%20F%20%5C%5C%0AE%2C%20%26%20C%2C%20%26%20D%2C%20%26%20F%20%5C%5C%0A%5Cend%7Bbmatrix%7D%0A" alt="
+\begin{bmatrix}
+E, &amp; O, &amp; F
+\end{bmatrix} = 
+\begin{bmatrix}
+E, &amp; A, &amp; B, &amp; F \\
+E, &amp; C, &amp; D, &amp; F \\
+\end{bmatrix}
+" />
+
+<img src="http://tex.s2cms.ru/svg/%0A%5Cbegin%7Bbmatrix%7D%0AO%2C%20%26%20P%0A%5Cend%7Bbmatrix%7D%20%3D%20%0A%5Cbegin%7Bbmatrix%7D%0AA%2C%20%26%20B%2C%20%26%20E%2C%20%26%20F%20%5C%5C%0AA%2C%20%26%20B%2C%20%26%20G%2C%20%26%20H%20%5C%5C%0AC%2C%20%26%20D%2C%20%26%20E%2C%20%26%20F%20%5C%5C%0AC%2C%20%26%20D%2C%20%26%20G%2C%20%26%20H%20%5C%5C%0A%5Cend%7Bbmatrix%7D%0A" alt="
+\begin{bmatrix}
+O, &amp; P
+\end{bmatrix} = 
+\begin{bmatrix}
+A, &amp; B, &amp; E, &amp; F \\
+A, &amp; B, &amp; G, &amp; H \\
+C, &amp; D, &amp; E, &amp; F \\
+C, &amp; D, &amp; G, &amp; H \\
+\end{bmatrix}
+" />
+
+Note that buildpack IDs are expanded depth-first in left-to-right order.
+
+If a buildpack entry within a group has the parameter `optional = true`, the a copy of the group without the entry is repeated after the original group. This is functionally equivalent to the current meaning of optional for a single buildpack implementation. The current meaning of optional for a single buildpack implementation will remain unchanged, as it is more performant and results in the same behavior.
+
+
+## User Interface
+
+### App Developer
+
+`pack build` should accept a list of buildpacks via the `--buildpack` flag.
+The flag may be passed multiple times to construct a buildpack group.
+
+The value of each flag must be one of:
+- A buildpack ID of a buildpack on the builder, optionally followed by `@` and a version
+- A path to a buildpackage on the local filesystem
+- A reference to a buildpackage on a Docker registry  
```

Comment:
In our development workflow this seems non-ideal, because it looks like we lose the ability to directly use a buildpack, because we have to now go:
`buildpack-code -> [compiled buildpack] -> [buildpackage]. `
That makes development of buildpacks a bit more cumbersome, I think.

Reply:
That’s a good point. Ideally, the whole compilation process could be wrapped into `pack create-package`, but we would definitely need some kind of pre-package hook in `buildpack.toml`.

Reply:
I also do local buildpack development this way with `--buildpack` using `pack`. There's a huge ease of use factor when being able to use an exploded buildpack directory.